README
NAME
    PerlX::Window - sliding windows on a string or array

SYNOPSIS
       use feature qw(say);
       use PerlX::Window;
   
       my $string = "Foobar";
   
       while (defined window $string, 3)
       {
          say $window;  # says "Foo"
                        # says "oob"
                        # says "oba"
                        # says "bar"
       }

DESCRIPTION
    This module provides a sliding window over a long string or array. It
    exports two functions `window` and `window_pos`, and two variables $window
    and @window.

    `window $string, $length`
        Calling this function returns the current window onto the string, and
        increments the stored position. The window returned is an *lvalue*
        which means you can assign to it (like `substr`).

        Once the string has been exhausted, it returns `undef` (or in list
        context, the empty list), and resets the stored position for the
        string.

    `window @array, $length`
        Like the string version, but instead of operating on a substring of a
        string, operates on a slice of an array.

    `window_pos $string`
        Returns the position of the most recent window onto the string; a
        zero-indexed integer.

    `window_pos @array`
        Returns the position of the most recent window onto the array; a
        zero-indexed integer.

    `window_pos`
        Called with no arguments, defaults to the string or array from the
        most recent call to `window`.

    $window
        An alias to the current window onto the string that has most recently
        had `window` called upon it.

        $window is implemented using Variable::Magic if installed, and a tie
        otherwise.

    @window
        An alias to the current window onto the array that has most recently
        had `window` called upon it.

        You may not assign to this in list context, nor perform `pop`, `push`,
        `shift`, `unshift`, or `slice` operations on it, nor any other
        operation that would change the length of the array. You may however
        assign to indexes within the array:

           $window[0] = "Fee" if $window[0] eq "Foo";

        @window is implemented using a tie.

CAVEATS
    `window` is prototyped `(\[$@]$)` which means that the first argument must
    be a literal scalar or array variable, and `window` will actually fetch a
    reference to that variable. This means the following are not the same:

       my $tmp = "Foobar";
       say $window
          while window $tmp, 3;

       say $window
          while window my $tmp = "Foobar", 3;

    The second example says "Foo" infinitely because $tmp is redefined in each
    loop, so is a separate variable as far as `window` is concerned.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=PerlX-Window>.

SEE ALSO
    Data::Iterator::SlidingWindow.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
PerlX-Window
============

Created:      2014-04-28
Home page:    <https://metacpan.org/release/PerlX-Window>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=PerlX-Window>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.003	2014-09-16

 [ Packaging ]
 - Add repo link.

0.002	2014-05-18

 [ Packaging ]
 - Requires Perl 5.16.
   Fixes RT#95741.
   Slaven Rezic++
   <https://rt.cpan.org/Ticket/Display.html?id=95741>

0.001	2014-04-28	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "a module that does something-or-other",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 1,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.021, CPAN::Meta::Converter version 2.140640",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "PerlX-Window",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       recommends => { "Hash::FieldHash" => 0, "Variable::Magic" => 0 },
                                       requires   => { "Data::Alias" => 0, "Exporter::Shiny" => 0, "perl" => "5.016000" },
                                     },
                        test      => { requires => { "Test::More" => 0.96 } },
                      },
  "provides"       => {
                        "PerlX::Window" => { file => "lib/PerlX/Window.pm", version => 0.003 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=PerlX-Window",
                                        },
                        homepage     => "https://metacpan.org/release/PerlX-Window",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-perlx-window.git",
                                          web  => "https://github.com/tobyink/p5-perlx-window",
                                        },
                        x_identifier => "http://purl.org/NET/cpan-uri/dist/PerlX-Window/project",
                      },
  "version"        => 0.003,
};

my %dynamic_config;
do {
# Hash::Util::FieldHash was first released with perl v5.9.4
$meta->{prereqs}{runtime}{requires}{'Hash::FieldHash'} = 0 if $] < 5.009004;

};

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

